src/lib/containers/table/datasets/DatasetItemsEditor.tsx

```diff
@@ -162,60 +182,91 @@ export function DatasetItemsEditor(props: DatasetItemsEditorProps) {
+              old.versionNumber !== result.versionNumber,
+          )
+        ) {
+          results['updatedItems'].push(result)
+        }
+        if (!oldDataSet.find(old => old.entityId === result.entityId)) {
+          results['newItems'].push(result)
+        }
+        if (
+          oldDataSet.find(
+            old =>
+              old.entityId === result.entityId &&
+              old.versionNumber === result.versionNumber,
+          )
+        ) {
+          unchangedItems.push(result)
+        }
+        return results
+      },
+      { updatedItems: [], newItems: [] } as {
+        updatedItems: DatasetItem[]
+        newItems: DatasetItem[]
+      },
+    )
+
+    return { unchangedItems, updatedItems, newItems, deletedItems }
+  }
+
+  function getToastMessageTitle() {
+    const { updatedItems, newItems, deletedItems } = getDataSetDifference(
+      previousDatasetToUpdate?.items!,
+      datasetToUpdate?.items!,
+    )
+    let toastTitle = ''
+
+    // "X items(s) deleted"
+    if (deletedItems.length > 0) {
+      toastTitle += `${deletedItems.length} Item${
+        deletedItems.length === 1 ? '' : 's'
+      } removed from Dataset`
+    } else {
+      // "Y item(s) added"
+      toastTitle += `${newItems.length} Item${
+        newItems.length === 1 ? '' : 's'
+      } added`
+
+      // "and Z item(s) updated", only shown if there are updated items
+      if (updatedItems.length > 0) {
+        toastTitle += ` and ${updatedItems.length} Item${
+          updatedItems.length === 1 ? '' : 's'
+        } updated`
+      }
+      toastTitle += ` to Dataset`
+    }
+    return toastTitle
+  }
+
   function addItemsToDataset(itemsToAdd: Reference[]) {
     setDatasetToUpdate(datasetToUpdate => {
       if (datasetToUpdate) {
-        // Items that were already in the dataset and are not being updated
-        const unchangedItems = datasetToUpdate.items.filter(
-          item =>
-            !itemsToAdd.find(newItem => newItem.targetId === item.entityId),
-        )
-
-        // Items that were already in the dataset, but were selected so the version may have been updated
-        const updatedItems = itemsToAdd.filter(newItem =>
-          datasetToUpdate.items.find(
-            existingItem => existingItem.entityId === newItem.targetId,
-          ),
-        )
-
-        // Items that were not previously in the dataset
-        const newItems = itemsToAdd.filter(
-          newItem =>
-            !datasetToUpdate.items.find(
-              existingItem => existingItem.entityId === newItem.targetId,
-            ),
+        const refToDatasetItem = itemsToAdd.map(item => ({
+          entityId: item.targetId,
+          versionNumber: item.targetVersionNumber!,
+        }))
+        const { unchangedItems, updatedItems, newItems } = getDataSetDifference(
+          datasetToUpdate.items,
+          refToDatasetItem,
         )
+        const items = [...unchangedItems, ...updatedItems, ...newItems]
```

Comment:
I see, we do need items, rather than just counts, because this was refactored to use the same logic 👍